File: NutriSmart/app/src/main/java/com/example/nutrismart/ui/screens/weather/WeatherViewModel.kt
package com.example.nutrismart.ui.screens.weather

import android.content.Context
import android.content.Intent
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class WeatherUiState(
    val isLoading: Boolean = true,
    val temperature: Int = 0,
    val condition: String = "",
    val humidity: Int = 0,
    val cityName: String = "Locating...",
    val recommendations: List<String> = emptyList()
)

class WeatherViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(WeatherUiState())
    val uiState: StateFlow<WeatherUiState> = _uiState.asStateFlow()

    fun fetchRealWeather(lat: Double, lon: Double) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            try {
                val response = WeatherRetrofit.api.getCurrentWeather(lat, lon)

                if (response.isSuccessful) {
                    val body = response.body()
                    if (body != null) {
                        val currentTemp = body.main.temp.toInt()
                        val currentCondition = body.weather.firstOrNull()?.main ?: "Unknown"
                        val currentHumidity = body.main.humidity

                        val recs = generateRecommendations(currentTemp, currentCondition)

                        _uiState.update {
                            it.copy(
                                isLoading = false,
                                temperature = currentTemp,
                                condition = currentCondition,
                                humidity = currentHumidity,
                                cityName = body.name,
                                recommendations = recs
                            )
                        }
                    }
                } else {
                    _uiState.update { it.copy(isLoading = false, condition = "API Error", cityName = "Error") }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                _uiState.update { it.copy(isLoading = false, condition = "No Connection", cityName = "Offline") }
            }
        }
    }

    private fun generateRecommendations(temp: Int, condition: String): List<String> {
        val recs = mutableListOf<String>()

        if (temp >= 30) {
            recs.add("It's a heatwave! Make sure to increase your daily water intake.")
            recs.add("Avoid direct sun exposure between 11:00 AM and 4:00 PM.")
        }

        if (condition.contains("Rain", ignoreCase = true) || condition.contains("Drizzle", ignoreCase = true)) {
            recs.add("It's raining outside. Consider replacing your outdoor run with an indoor workout.")
        } else if (temp in 15..25 && !condition.contains("Rain", ignoreCase = true)) {
            recs.add("Perfect weather for a walk or an outdoor run!")
        }

        return recs
    }

    fun checkHydrationAndNotify(context: Context, waterConsumedMl: Int) {
        val state = _uiState.value
        if (state.temperature >= 30 && waterConsumedMl < 1000) {
            val intent = Intent(context, WeatherNotificationReceiver::class.java)
            context.sendBroadcast(intent)
        }
    }
}

File: NutriSmartIOS/ui/screens/weather/WeatherViewModel.swift
//
//  WeatherViewModel.swift
//  NutriSmartIOS
//

import Foundation
import SwiftUI
import CoreLocation

struct WeatherUiState {
    var isLoading = true
    var temperature: Int = 0
    var condition: String = ""
    var humidity: Int = 0
    var cityName: String = "Locating..."
    var recommendations: [String] = []
}

@Observable
@MainActor
class WeatherViewModel {
    var uiState = WeatherUiState()
    private let weatherRepository = WeatherRepository.shared
    
    func fetchRealWeather(lat: Double, lon: Double) {
        uiState.isLoading = true
        Task {
            do {
                let response = try await weatherRepository.getCurrentWeather(lat: lat, lon: lon)
                let currentTemp = Int(response.main.temp)
                let currentCondition = response.weather.first?.main ?? "Unknown"
                let recs = generateRecommendations(temp: currentTemp, condition: currentCondition)
                
                uiState.temperature = currentTemp
                uiState.condition = currentCondition
                uiState.humidity = response.main.humidity
                uiState.cityName = response.name
                uiState.recommendations = recs
                uiState.isLoading = false
            } catch {
                uiState.condition = "Error"
                uiState.cityName = "Error"
                uiState.isLoading = false
            }
        }
    }
    
    private func generateRecommendations(temp: Int, condition: String) -> [String] {
        var recs: [String] = []
        if temp >= 30 {
            recs.append("It's a heatwave! Increase water intake.")
            recs.append("Avoid direct sun 11AM-4PM.")
        }
        if condition.lowercased().contains("rain") {
            recs.append("It's raining. Indoor workout?")
        } else if temp >= 15 && temp <= 29 {
            recs.append("Perfect weather for a walk!")
        } else {
            recs.append("Go out for a walk, but dress well.")
        }
        return recs
    }
}
